orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+                    },
+                    "optimizer":
+                    {
+                        model_weight_name:
+                        {
+                            "Moment_1": moment1_tensor,
+                            "Moment_2": moment2_tensor,
+                            "Update_Count": update_tensor # if optimizer is adam, absent otherwise
+                        },
+                        "shared_optimizer_state": # if optimizer is shared, absent otherwise.
+                        {
+                            "step": step_tensor # int array of size 1
+                        }
+                    }
+                    "trainer_options":
+                    {
+                        "mixed_precision": mixed_precision_enabled_bool,
+                        "zero_stage": zero_enabled_stage,
+                        "world_rank": world_rank_int,
+                        "world_size": world_size_int,
+                        "optimizer_name": training_optimizer_name
+                    },
+                    "partition_info":
+                    {
+                        model_weight_name:
+                        {
+                            "original_dim": original_dimension_array
+                        }
+                    }
+                }
+            - when pytorch_format is true
+                {
+                    model_weight_name: model_weight_tensor
+                }
+
+        """
+        if not self._training_session:
+            warnings.warn("ONNX Runtime training session is not initialized yet. " \
+                          "Please run train_step or eval_step at least once before calling ORTTrainer.state_dict().", \
+                          UserWarning)
+            return self._load_state_dict.args[0] if self._load_state_dict else {}
+
+        states = {}
+
+        # load training session model states into the states dictionary
+        self._extract_model_states(states)
+        if pytorch_format:
+            # if pytorch_format is true, return a flat dictionary with only model states
+            # which is compatible with a PyTorch model
+            return states['model']['fp32']
+
+        # load training session optimizer states into the states dictionary
+        self._extract_optimizer_states(states)
+
+        # extract the relevant training configuration from the trainer and load them into the states dictionary
+        self._extract_trainer_options(states)
+
+        # add partition information in case of a distributed run
+        if self.options.distributed.deepspeed_zero_optimization.stage > 0:
+            partition_info = 'partition_info'
+            states[partition_info] = self._training_session.get_partition_info_map()
+
+        return states
+
+    def _load_model_states(self, training_session_state_dict, state_dict, strict):
+        """Load the model states onto the training session state dictionary"""
+
+        model = 'model'
```

Comment:
can we define the model and optimizer keys in a single place so that any future change has to happen only once?

Reply:
Will do that.